BaikalDataGenerator/Source/render.cpp

```diff
@@ -0,0 +1,437 @@
+/**********************************************************************
+Copyright (c) 2016 Advanced Micro Devices, Inc. All rights reserved.
+
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in
+all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+********************************************************************/
+
+#include "render.h"
+#include "scene_io.h"
+#include "material_io.h"
+#include "SceneGraph/light.h"
+#include "Output/clwoutput.h"
+#include "BaikalIO/image_io.h"
+
+#include <fstream>
+#include "XML/tinyxml2.h"
+
+using namespace Baikal;
+
+static std::uint32_t constexpr kNumIterations = 4096;
+
+struct OutputDesc
+{
+    Renderer::OutputType type;
+    std::string name;
+    // extansion of the file to save
+    std::string file_ext;
+    int bpp;
+};
+
+// if you need to add new output for saving to disk
+// just put its description in thic collection
+std::vector<OutputDesc> outputs_collection = { { Renderer::OutputType::kColor, "color", "png", 16 },
+                                               { Renderer::OutputType::kViewShadingNormal, "view_shading_normal", "png", 8 },
+                                               { Renderer::OutputType::kDepth, "depth", "png", 16 },
+                                               { Renderer::OutputType::kAlbedo, "albedo", "png", 8 },
+                                               { Renderer::OutputType::kGloss, "gloss", "png", 8 } };
+
+static bool operator != (RadeonRays::float3 left, RadeonRays::float3 right)
+{
+    return (left.x != right.x) ||
+        (left.y != right.y) ||
+        (left.z != right.z);
+}
+
+Render::Render(const std::string &file_name,
+               const std::string &path,
+               std::uint32_t output_width,
+               std::uint32_t output_height)
+{
+    std::vector<CLWPlatform> platforms;
+    CLWPlatform::CreateAllPlatforms(platforms);
+
+    int platform_index = 0;
+
+    for (auto j = 0u; j < platforms.size(); ++j)
```

Comment:
range-based for ?

Reply:
Yes, or std::find_if as suggested below, depending on which variant is more compact/readable.

Reply:
removed